zebra-network/src/peer/priority.rs

```diff
@@ -0,0 +1,116 @@
+///
+/// Outbound peer connections are initiated in the sorted [order](std::ops::Ord) of this type.
+///
+/// In particular, public addresses and non-canonical ports are preferred to
+/// private addresses and canonical ports.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Ord, PartialOrd)]
+pub struct PeerPreference {
+    /// Is the peer using the canonical Zcash port for the configured [`Network`]?
+    canonical_port: AttributePreference,
+}
+
+impl AttributePreference {
+    /// Returns `Preferred` if `is_preferred` is true.
+    pub fn preferred_from(is_preferred: bool) -> Self {
+        if is_preferred {
+            Preferred
+        } else {
+            Acceptable
+        }
+    }
+
+    /// Returns `true` for `Preferred` attributes.
+    #[allow(dead_code)]
+    pub fn is_preferred(&self) -> bool {
+        match self {
+            Preferred => true,
+            Acceptable => false,
+        }
+    }
+}
+
+impl PeerPreference {
+    /// Return a preference for the peer at `peer_addr` on `network`.
+    ///
+    /// Use the [`PeerPreference`] [`Ord`] implementation to sort preferred peers first.
+    pub fn new(
+        peer_addr: &SocketAddr,
+        network: impl Into<Option<Network>>,
+    ) -> Result<PeerPreference, &'static str> {
+        address_is_valid_for_outbound_connections(peer_addr, network)?;
+
+        // This check only prefers the configured network,
+        // because the address book and initial peer connections reject the port used by the other network.
+        let canonical_port =
+            AttributePreference::preferred_from([8232, 18232].contains(&peer_addr.port()));
+
+        Ok(PeerPreference { canonical_port })
+    }
+}
+
+/// Is the [`SocketAddr`] we have for this peer valid for outbound
+/// connections?
+///
+/// Since the addresses in the address book are unique, this check can be
+/// used to permanently reject entire [`MetaAddr`]s.
+fn address_is_valid_for_outbound_connections(
+    peer_addr: &SocketAddr,
+    network: impl Into<Option<Network>>,
+) -> Result<(), &'static str> {
+    // TODO: make private IP addresses an error unless a debug config is set (#3117)
+
+    if peer_addr.ip().is_unspecified() {
+        return Err("invalid peer IP address: unspecified addresses can not be used for outbound connections");
+    }
+
+    // 0 is an invalid outbound port.
+    if peer_addr.port() == 0 {
+        return Err(
+            "invalid peer port: unspecified ports can not be used for outbound connections",
+        );
+    }
+
+    // Ignore ports used by similar networks: Flux/ZelCash and misconfigured Zcash.
+    if let Some(network) = network.into() {
+        if peer_addr.port() == network.default_port() {
+            return Ok(());
+        }
+
+        if peer_addr.port() == 8232 {
+            return Err(
+                "invalid peer port: port is for Mainnet, but this node is configured for Testnet",
+            );
+        } else if peer_addr.port() == 18232 {
```

Comment:
TODO: add 18344 regtest case here:
https://github.com/ZcashFoundation/dnsseeder/blob/8aa4e10c8fbfdfba8af00957ef5d49717432c215/zcash/network/params.go#L14

Reply:
Fixed in [9448698](https://github.com/ZcashFoundation/zebra/pull/4564/commits/9448698bafdf22447318bfa8a0ce285eaef909f8)